src/postamble_minimal.js

```diff
@@ -25,6 +25,11 @@ function run() {
 
 #if EXIT_RUNTIME
   callRuntimeCallbacks(__ATEXIT__);
+#endif
+#if USE_PTHREADS
+  PThread.terminateAllThreads();
+#endif
```

Comment:
I think the per-thread exist handler should go up before the program-wide exit handlers.. (so up on line 25).   If you do that perhaps we can avoid the lsan changes?  I'm not sure.

So maybe the overall order should be:

1. per-thread-exit handler for this thread (`PThread.runExitHandlers();`)
2. kill all other threads (PThread.terminateAllThreads())
3. program-wide exit handlercallRuntimeCallbacks(__ATEXIT__);s (callRuntimeCallbacks(__ATEXIT__);
4. program-wide JS exit handlers (<<< ATEXITS >>>)

Reply:
`PThread.terminateAllThreads()` (2) needs to be run after the program-wide exit handlers (3) and possibly also after the program-wide JS exit handlers (4). To test this, see the added `test_pthread_atexit` in this PR which deadlocks on the main branch.

I've revised the order to:
1. program-wide exit handlers (`callRuntimeCallbacks(__ATEXIT__)`).
2. program-wide JS exit handlers (`  <<< ATEXITS >>>`).
3. terminate all other threads (`PThread.terminateAllThreads()`).
4. call exit handlers for the main thread (`PThread.runExitHandlers();`).
   (this could probably be removed after PR #14489, but I'm not sure)

I've reverted the LSan changes for now, and just suppressed `_emscripten_builtin_pthread_create` from being leaky with commit https://github.com/emscripten-core/emscripten/pull/14481/commits/95e23faed1614e96e2a6f04e69728ef70617afdf.

Reply:
> `PThread.terminateAllThreads()` (2) needs to be run after the program-wide exit handlers (3) and possibly also after the program-wide JS exit handlers (4). To test this, see the added `test_pthread_atexit` in this PR which deadlocks on the main branch.
> 

I still need to take a look at that test but this seems logically incorrect to me.   The program wide exit handler run, for example, the C++ static destructors.  If we don't kill all the threads before running these then we could have situation where there are threads still running with access to static C++ objects which have been destroyed.   For example imagine std::cout has a static destructor and one of the background threads tries to write to cout during this time?


> I've revised the order to:
> 
> 1. program-wide exit handlers (`callRuntimeCallbacks(__ATEXIT__)`).
> 2. program-wide JS exit handlers (`  <<< ATEXITS >>>`).
> 3. terminate all other threads (`PThread.terminateAllThreads()`).
> 4. call exit handlers for the main thread (`PThread.runExitHandlers();`).
>    (this could probably be removed after PR #14489, but I'm not sure)
> 
> I've reverted the LSan changes for now, and just suppressed `_emscripten_builtin_pthread_create` from being leaky with commit [95e23fa](https://github.com/emscripten-core/emscripten/commit/95e23faed1614e96e2a6f04e69728ef70617afdf).

Reply:
Actually I think maybe you are right.   Looking at the musl code for exit() is currently does:

```
_Noreturn void exit(int code)                                                    
{                                                                                
  __funcs_on_exit();                                                             
  __libc_exit_fini();                                                            
  __stdio_exit();                                                                
  _Exit(code)                                                                                                                  
}     
```

I think that secondary threads are not killed until `_Exit` is called (which calls `__syscall(SYS_exit_group, ec)` which I think terminates all the threads).

I guess that static destruction *can* run while there still threads running.

Reply:
Yup .. i ran quick test :

```
class Foo {                                                                      
  public:                                                                        
  Foo() {                                                                        
    printf("Foo\n");                                                             
  }                                                                              
  ~Foo() {                                                                       
    printf("~Foo\n");                                                            
    alive = false;                                                               
  }                                                                              
  bool alive = true;                                                             
};                                                                               
                                                                                 
static Foo foo;                                                                  
                                                                                 
static _Atomic int threadCounter = 0;                                            
static _Atomic int running = 1;                                                  
static pthread_t thread;                                                         
                                                                                 
void *workerThread(void* arg) {                                                  
  threadCounter++;                                                               
                                                                                 
  while (running) {                                                              
    printf("foo: %d\n", foo.alive);                                              
  }                                                                              
                                                                                 
  threadCounter--;                                                               
                                                                                 
  return NULL;                                                                   
}    
                                                                                    
void mainAtExit() {                                                                 
  printf("mainAtExit\n");                                                           
  sleep(1);                                                                         
}                                                                                   
                                                                                                                               
int main(int argc, char* argv[]) {                                                  
  int rc = atexit(mainAtExit);                                                      
  assert(rc == 0);                                                                  
                                                                                    
  rc = pthread_create(&thread, NULL, workerThread, NULL);                           
  assert(rc == 0);                                                                  
  return 0;                                                                         
}     
```

Reply:
Output ends with:

```
foo: 1
foo: 1
foo: 1
foo: 1
foo: 1
~Foo
foo: 1
foo: 0
foo: 0
foo: 0
foo: 0
foo: 0
foo: 0
foo: 0
foo: 0
foo: 0
foo: 0
foo: 0
foo: 0
foo: 0
foo: 0
foo: 0
foo: 0
```

Reply:
Do you are correct terminateAllThreads should come last.. and threads that are still alive that point might have a bad time but that is to be expected.

Reply:
Thanks for the investigation. Given that `PThread.terminateAllThreads()` isn't last now, do you think `PThread.runExitHandlers()` should occur before `PThread.terminateAllThreads()` (i.e. swapping 3 and 4 in the above order)?

Also, let me know if you prefer commit https://github.com/emscripten-core/emscripten/pull/14481/commits/58054658401169a67084c7e97040e07cdf5bcd71 as a separate PR.

Reply:
Yes, that would make a good separate PR I think.  IIUC that is not a file that exists upstream right?

Reply:
`terminateAllThreads()` should come last of all I think.

Reply:
`asan_emscripten.cpp` does indeed not exist upstream. I've moved that commit to PR #14516, and moved `PThread.terminateAllThreads()` to occur as last.